src/pretix/plugins/socialauth/views.py

```diff
@@ -18,43 +20,107 @@
 from pretix.helpers.urls import build_absolute_uri
 
 from .schemas.login_providers import LoginProviders
+from .schemas.oauth2_params import OAuth2Params
 
 logger = logging.getLogger(__name__)
 adapter = get_adapter()
 
 
-def oauth_login(request, provider):
-    gs = GlobalSettingsObject()
-    client_id = gs.settings.get('login_providers', as_type=dict).get(provider, {}).get('client_id')
-    provider = adapter.get_provider(request, provider, client_id=client_id)
-
-    base_url = provider.get_login_url(request)
-    query_params = {
-        "next": build_absolute_uri("plugins:socialauth:social.oauth.return")
-    }
-    parsed_url = urlparse(base_url)
-    updated_url = parsed_url._replace(query=urlencode(query_params))
-    return redirect(urlunparse(updated_url))
+class OAuthLoginView(View):
+    def get(self, request: HttpRequest, provider: str) -> HttpResponse:
+        self.set_oauth2_params(request)
 
+        gs = GlobalSettingsObject()
+        client_id = (
+            gs.settings.get("login_providers", as_type=dict)
+            .get(provider, {})
+            .get("client_id")
+        )
+        provider_instance = adapter.get_provider(request, provider, client_id=client_id)
+
+        base_url = provider_instance.get_login_url(request)
+        query_params = {
+            "next": build_absolute_uri("plugins:socialauth:social.oauth.return")
+        }
+        parsed_url = urlparse(base_url)
+        updated_url = parsed_url._replace(query=urlencode(query_params))
+        return redirect(urlunparse(updated_url))
+
+    @staticmethod
+    def set_oauth2_params(request: HttpRequest) -> None:
+        """
+        Handle Login with SSO button from other components
+        This function will set 'oauth2_params' in session for oauth2_callback
+        """
+        next_url = request.GET.get("next", "")
+        if not next_url:
+            return
+
+        parsed = urlparse(next_url)
+
+        # Only allow relative URLs
+        if parsed.netloc or parsed.scheme:
+            return
+
+        params = parse_qs(parsed.query)
+        sanitized_params = {
+            k: v[0].strip()
+            for k, v in params.items()
+            if k in OAuth2Params.model_fields.keys()
+        }
+
+        try:
+            oauth2_params = OAuth2Params.model_validate(sanitized_params)
+            request.session["oauth2_params"] = oauth2_params.model_dump()
+        except ValidationError as e:
+            logger.error("Error while validating OAuth2 parameters: %s", e)
+
+
+class OAuthReturnView(View):
+    def get(self, request: HttpRequest) -> HttpResponse:
+        try:
+            user = self.get_or_create_user(request)
+            response = process_login_and_set_cookie(request, user, False)
+            oauth2_params = request.session.pop("oauth2_params", {})
+            if oauth2_params and self.validate_oauth2_params(oauth2_params):
```

Comment:
Now, you already adopt Pydantic model, you should use its validation result:

```py
try:
    oauth2_params = OAuth2Params.model_validate(oauth2_params_raw)
    auth_url = reverse("control:oauth2_provider.authorize")
    query_string = urlencode(oauth2_params.model_dump())
    return redirect(f"{auth_url}?{query_string}")
except ValidationError as e:
    logger.warning("Ignore invalid OAuth2 parameters: %s.", e)
return response
```

Here I change log level to `warning` because this issue of invalid OAuth params is not severe, we can continue to serve user without it.